hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -181,6 +181,13 @@ def __stoppable_test_function(self, data):
                     # need to re-raise it so that it will eventually reach the
                     # correct engine.
                     raise
+            except RecursionError:
+                # We treat RecursionError specially, and hint the user, since
+                # it may occur inside Hypothesis for otherwise completely valid strategies.
+                raise NestingTooDeep(
+                    f"Max recursion depth was exceeded during test execution. This is "
+                    "probably due to too deep strategy nesting."
+                )
```

Comment:
I'm pretty dubious about this - do we have concrete examples where this is a problem, and reason to think that there are few other cases it impacts?  Can we scope it more tightly, e.g. wrapped around drawing only?  

I'd also prefer to have it inherit from `RecursionError`, so that we don't break any existing try/except logic.